### DISCUSSION TITLE
DNSCrypt-proxy uses cloudflare as a DNS even when not specified in config file.

### DISCUSSION THREAD

**Discussion by kalpaj12** [DNSCrypt/dnscrypt-proxy#1742] [Q&A] [answered]
## Output of the following commands:

./dnscrypt-proxy -version 
`2.0.45`

./dnscrypt-proxy -check
```bash
[2021-06-11 07:44:10] [NOTICE] dnscrypt-proxy 2.0.45
[2021-06-11 07:44:10] [NOTICE] Source [public-resolvers] loaded
[2021-06-11 07:44:10] [NOTICE] Source [relays] loaded
[2021-06-11 07:44:11] [NOTICE] Configuration successfully checked
```

./dnscrypt-proxy -resolve example.com
```bash
Resolving [example.com] using 127.0.0.1 port 54

Resolver      : 74.63.20.246 (res101.qpg.rrdns.pch.net.)
Lying         : no
DNSSEC        : yes, the resolver supports DNSSEC

Canonical name: example.com.

IPv4 addresses: 93.184.216.34
IPv6 addresses: 2606:2800:220:1:248:1893:25c8:1946

Name servers  : b.iana-servers.net., a.iana-servers.net.
DNSSEC signed : yes
Mail servers  : 1 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : v=spf1 -all, 8j5nfqld20zpcyr8xjw0ydcfq9rk8hgm
```

## *What* is affected by this bug?
* I have Pi-hole installed on a Raspberry pi 4B. I am using a Custom upstream DNS (127.0.0.1#54). (**Note:** This is the only DNS Pi-hole send queries to for resolution)
* At port 54 of the Pi, is where I have dnscrypt-proxy listening at, with `server_names = ['quad9-doh-ip4-filter-pri']`
* [dnsleaktest](https://www.dnsleaktest.com/) shows that I resolve via WoodyNet (as expected for quad9), Cloudflare **(why?)**

## *When* does this occur?
Everytime! Flushed DNS, still occurs.

## *Where* does it happen?
-

## *How* do we replicate the issue?

-

## Expected behavior (i.e. solution)
No cloudflare DNS;-;

**Comment by kalpaj12** [comment]
My [**dnscrypt-proxy.toml**](https://paste.ubuntu.com/p/4zFzTdJmSf/)

**Comment by jedisct1** [comment]
> Resolver      : 74.63.20.246 (res101.qpg.rrdns.pch.net.)

pch.net is Quad9, not Cloudflare.

**Comment by kalpaj12** [comment]
@jedisct1, example2.com resolved via cloudflare's DNS, correct?

```bash
Resolving [example2.com] using 127.0.0.1 port 54

Resolver      : 162.158.234.59
Lying         : no
DNSSEC        : yes, the resolver supports DNSSEC

Canonical name: example2.com.

IPv4 addresses: 173.231.210.103
IPv6 addresses: -

Name servers  : ns1.inmotionhosting.com., ns2.inmotionhosting.com.
DNSSEC signed : no
Mail servers  : 1 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : -
```

Also, I tried example.com again
```bash
Resolving [example.com] using 127.0.0.1 port 54

Resolver      : 162.158.234.59
Lying         : no
DNSSEC        : yes, the resolver supports DNSSEC

Canonical name: example.com.

IPv4 addresses: 93.184.216.34
IPv6 addresses: 2606:2800:220:1:248:1893:25c8:1946

Name servers  : b.iana-servers.net., a.iana-servers.net.
DNSSEC signed : yes
Mail servers  : 1 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : v=spf1 -all, 8j5nfqld20zpcyr8xjw0ydcfq9rk8hgm
```

**Comment by jedisct1** [comment]
If you stop the proxy, what does dnsleaktest show?

**Reply by kalpaj12** [reply]
@jedisct1 I stopped the proxy, and flushed DNS, couldn't load any website. Had to start the proxy again.

**Reply by jedisct1** [reply]
Can you try to start the proxy with random junk in the configuration file to verify that this is the file actually being loaded?

**Comment by kalpaj12** [comment, marked as answer]
I added [Quad9 resolvers](https://www.quad9.net/service/service-addresses-and-features#dnscrypt) to [sources] of the configuration file, and I guess that did the job! I now have only Quad9 DNS's in dnsleaktest.

This is my current configuration file [dnscrypt-proxy.toml](https://paste.ubuntu.com/p/vsPQsWPn4V/). This works!

P.s. found out Pi-hole is better with using Unbound as a recursive DNS. Although that, but great work team dnscrypt-proxy! :)

**Comment by lifenjoiner** [comment]
I think the strange part is the resolver IP.

I did tests on Windows with:
1. Only 1 server is specified by server_names. I used `uad9-dnscrypt-ip4-filter-pri`.
2. `file = '/dev/stdout'`
3. Run the server in a `cmd`
4. Run `Wrieshark`
5. Run `dnscrypt-proxy -resolve example.com` in another `cmd`

Similar results. For example:
```
Resolver      : 74.63.25.248 (res110.ams.rrdns.pch.net.)
```
In the server window, `quad9-dnscrypt-ip4-filter-pri` is used for all queries can be seen. Good.
But filtering Wrieshark results, you can see just `9.9.9.9` or `149.112.112.9` is used, rather than `74.63.25.248`.

https://github.com/DNSCrypt/dnscrypt-proxy/blob/a34258c3aa09f0382cc705d570038533124177bf/dnscrypt-proxy/resolve.go#L85-L104

**Reply by lifenjoiner** [reply]
Here `Resolver` is not the remote server. It is the builtin `resolver.dnscrypt.info`.
Try `dnscrypt-proxy -resolve resolver.dnscrypt.info`, you will see.
https://github.com/DNSCrypt/dnscrypt-proxy/blob/a34258c3aa09f0382cc705d570038533124177bf/dnscrypt-proxy/resolve.go#L74
